src/tensor.jl

```diff
@@ -28,188 +34,119 @@ ITensors.store(a::QXTensor) = a.storage
+    if !(storage isa MockTensor)
+        if diagonal_check && hyper_indices === nothing
+            hyper_indices = find_hyper_edges(storage)
+        end
+        # if it is not already in reduced form then we reduce it
+        if length(hyper_indices) > 0 && ndims(storage) == length(indices)
+            storage = reduce_tensor(storage, hyper_indices)
+        end
     end
-    QXTensor(indices, hyper_indices, data)
-end
 
-"""
-    Base.convert(::Type{ITensor}, t::QXTensor)
+    if hyper_indices === nothing
+        hyper_indices = Vector{Vector{Int64}}()
+    end
 
-Overloaded convert to enable conversion of a QXTensor to an ITensor. Note that doing this
-will mean that information about the hyper indices will be lost.
-"""
-function Base.convert(::Type{ITensor}, t::QXTensor)
-    ITensor(t.storage, IndexSet(t.indices))
+    QXTensor(length(indices), indices, hyper_indices, storage)
 end
 
-"""
-    Base.convert(::Type{QXTensor}, t::ITensor{N}) where N
-
-Overloaded convert to enable conversion of an ITensor to a QXTensor. Note that converting
-in this way means that hyper indices won't be identified.
-"""
-function Base.convert(::Type{QXTensor}, t::ITensor{N}) where N
-    QXTensor(store(t), collect(inds(t)))
-end
+QXTensor(i::Vector{<:Index}, t::Union{Nothing, <: AbstractArray}; kwargs...) =
+         QXTensor(i, nothing, t; kwargs...)
+QXTensor(t::Union{Nothing, <: AbstractArray}; kwargs...) =
+         QXTensor([Index(x) for x in size(t)], nothing, t; kwargs...)
 
 """
-    tensor_data(tensor::QXTensor; consider_hyperindices::Bool=false)
+    tensor_data(tensor::QXTensor; consider_hyperindices::Bool=true)
 
 Get the data associated with the given tensor. If the consider_hyperindices flag is true
-then only the first of the hyper indices are retained. For example for a 5 rank tensor where
+then the rank is reduced to merge related indices. For example for a 5 rank tensor where
 the 2nd and 4th indices form a group of hyper indices, with this option set to true would
-return a rank 4 tensor where the 2nd index
-"""
-function tensor_data(tensor::QXTensor; consider_hyperindices::Bool=false)
-    tensor_dims = Tuple([dim(x) for x in inds(tensor)])
-    data = reshape(convert(Array, store(tensor)), tensor_dims)
-    if consider_hyperindices
-        hi = hyperindices(tensor)
-        # create an array of the ranks from the groups of hyper indices
-        hi_ranks = Array{Int64, 1}[]
-        all_indices = inds(tensor)
-        for group in hi
-            push!(hi_ranks, map(x -> findfirst(y -> y == x, all_indices) ,group))
-        end
-        return reduce_tensor(data, hi_ranks)
-    else
+return a rank 4 tensor where the 2nd index. With hyperindices set to false a rank 5 tensor
+is returned.
+"""
+function tensor_data(tensor::QXTensor; consider_hyperindices::Bool=true)
+    data = store(tensor)
+    hi = hyperindices(tensor)
+    if consider_hyperindices || length(hi) == 0
```

Comment:
I'm not sure if I've followed the logic correctly here but it looks like this should be ```if !(consider_hyperindices) || length(hi) == 0```

Reply:
Data is stored internally in reduced for with local hyperindices taken into account (since this is more efficient than storing full tensor and there is no access to global hyperindex information) so this is saying that if we are considering hyperindices or there are no hyperindices then just return the data as is, otherwise we will need to expand it.